Automattic/jetpack
# AI Client: expose and use JWT library

On this PR:
* Adds and exposes a JWT library for the AI completion endpoint. 
* [Imports](https://github.com/Automattic/jetpack/pull/31924/files#diff-b220faa6b9ab83e6204db50283e09b04e59cb5c34a5ba74f37f2ca8650f3fe24R4) and [uses](https://github.com/Automattic/jetpack/pull/31924/files#diff-b220faa6b9ab83e6204db50283e09b04e59cb5c34a5ba74f37f2ca8650f3fe24R58) requestJwt() function in the Jetpack plugin to get the token.
* Add TS support
* Jetpack AI JWT library [doc file](https://github.com/Automattic/jetpack/blob/d55b002af78438ec520b5513022d36775a8fc129/projects/js-packages/ai-client/src/jwt/Readme.md) (added on this PR)

Fixes https://github.com/Automattic/jetpack/issues/31933
Fixes https://github.com/Automattic/jetpack/issues/31934

## Proposed changes:

* AI Client: expose and use the JWT library

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






Since the Jetpack plugin consumes the `requestJwt()` function, the best way to test these changes is by using the AI Assistant block.

* Go to the block editor
* Set the debug() scope

```
localStorage.setItem( 'debug', 'jetpack-*' )
```
* Use the AI Assistant to generate content.
* Confirm the app requests the token as usual, paying attention to the debug lines

<img width="522" alt="Screenshot 2023-07-18 at 11 04 30" src="https://github.com/Automattic/jetpack/assets/77539/7fdd715f-850b-43a7-8958-f7a8f8fa2344">

* Request new changes
* Confirm the app uses the stored token when the request is made in the next two minutes

<img width="567" alt="Screenshot 2023-07-18 at 11 05 57" src="https://github.com/Automattic/jetpack/assets/77539/bef9e464-4a5f-4cab-a527-44bc1493de24">

### Storage

Using the client tool (Application tab in Chrome), confirm the token data is appropriately stored

<img width="110" alt="image" src="https://github.com/Automattic/jetpack/assets/77539/5c770de1-50e1-4592-9a4c-b3b1c100e61f">

### Endpoints depending on site type

* Confirm for Jetpack and Atomic  site it hits the Jetpack endpoint 

<img width="841" alt="Screenshot 2023-07-18 at 11 14 58" src="https://github.com/Automattic/jetpack/assets/77539/e1a4f7ca-13de-47e2-bbd2-58992520f2e2">

* Confirm for Simple sites it hits the wpcom endpoint

<img width="831" alt="Screenshot 2023-07-18 at 11 44 48" src="https://github.com/Automattic/jetpack/assets/77539/b0523618-5e23-4bba-aead-68b7a2508849">
